### PR TITLE
Add support for 1.17

### PR DESCRIPTION
**Pull request by Archengius**

### CODE REVIEW

**Comment by u9g** [comment]
needs lint

**Comment by rom1504** [comment]
blocked by https://github.com/PrismarineJS/node-minecraft-protocol/pull/866

File: lib/plugins/entities.js
Comment (rom1504) [inline lib/plugins/entities.js:271]: in 1.17 it's now a single varint named entityId

File: lib/plugins/game.js
Comment (extremeheat) [inline lib/plugins/game.js:59]: can you inline these if they only get called once? It makes the code easier to read IMO

File: lib/plugins/blocks.js
Comment (extremeheat) [inline lib/plugins/blocks.js:99]: where does the protocol return BigInts?

Comment (Archengius) [reply lib/plugins/blocks.js:99]: In the primaryBitMask field of packet_map_chunk of 1.17 protocol. All chunk related mask fields are now serialized as long arrays representing bit sets in fact, including lighting packet ones. That's needed to support extended height.

File: lib/plugins/game.js
Comment (Archengius) [reply lib/plugins/game.js:59]: I could, but I literally moved them out into separate functions because not having version dependent logic in the common code actually makes code easier to understand.

File: lib/plugins/blocks.js
Comment (rom1504) [reply lib/plugins/blocks.js:99]: I think the format of chunk packet in protocol 1.17.json is wrong, we should change it so things like this are not required.

File: lib/plugins/blocks.js
Comment (rom1504) [inline lib/plugins/blocks.js:314]: Flatten should not be needed. If it is needed, then the format of chunk packet is wrong.

Comment (Archengius) [reply lib/plugins/blocks.js:314]: I don't really know what you mean by "wrong", because format of the packet is correct and matches what the game actually sends (and for that particular field, game sends varint length prefixed array of 64-bit integers)

Comment (Archengius) [reply lib/plugins/blocks.js:314]: Same applies to section mask in map_chunk

File: lib/plugins/blocks.js
Comment (extremeheat) [reply lib/plugins/blocks.js:99]: I looked through the Minecraft java source and it is writing it as a 64-bit big endian array, so I don't get what this flipping this is about.
```java
         var1.writeLongArray(this.storage.getRaw()); // PalettedContainer.java
```
```java
   public FriendlyByteBuf writeLongArray(long[] var1) { // FriendlyByteBuf.java
      this.writeVarInt(var1.length);
      long[] var2 = var1;
      int var3 = var1.length;

      for(int var4 = 0; var4 < var3; ++var4) {
         long var5 = var2[var4];
         this.writeLong(var5);
      }

      return this;
   }
   ...
   public ByteBuf writeLong(long var1) {
      return this.source.writeLong(var1);
   }

   public ByteBuf writeLongLE(long var1) {
      return this.source.writeLongLE(var1);
   }
```

File: lib/plugins/blocks.js
Comment (rom1504) [reply lib/plugins/blocks.js:314]: yes my point is that if we need to unparse what protodef just parsed at every time we use it, then the format we used is useless, and we should for example return a buffer and parse it directly in pchunk
mineflayer should not do any chunk parsing

File: lib/plugins/blocks.js
Comment (rom1504) [reply lib/plugins/blocks.js:99]: this should at worse be happening in pchunk, not here
at best be unneeded by improving the definition of the packet in protocol.json

Comment (Archengius) [reply lib/plugins/blocks.js:99]: Flipping is needed because BitSet implementation in prismarine-chunk requires least significant bits to come first pretty much because bit addressing starts from zero and grows upwards.
Original implementation in Java doesn't need to do this because it actually operates on 64-bit integers and not 32-bit ones, but in JS we're kinda forced to use 32-bit numbers.

File: lib/plugins/entities.js
Comment (u9g) [reply lib/plugins/entities.js]: re this: before this is merged, we should just emit a RemoveEntities packet with the one entity from the RemoveEntity packet because in 1.17.1, we get the RemoveEntities packet back

**Comment by rom1504** [comment]
pchunk now has 1.17, let's finish this

File: lib/plugins/blocks.js
Comment (rom1504) [reply lib/plugins/blocks.js:99]: solved this

File: lib/plugins/blocks.js
Comment (rom1504) [reply lib/plugins/blocks.js:314]: solved

**Comment by rom1504** [comment]
from @nickelpro : 
> On login, you get min_y and height from the login packet. These need to be passed prismarine-world (which needs to be updated) somehow, which then needs o pass them as minY and worldHeight in the options object to the ChunkColumn constructor.
> The defaults will work for worlds that don't mess with min_y and height, but Mojang has already released official world generator data packs that modify these

File: lib/plugins/game.js
Comment (rom1504) [reply lib/plugins/game.js:59]: this should be inlined indeed

**Review by malafk — APPROVED** [review]
loooks good

**Comment by rom1504** [comment]
>      Uncaught TypeError: Cannot read property 'min_y' of undefined
in 1.8

**Comment by extremeheat** [comment]
Should be fixed, what's failing in 1.17 currently:
* crafting
* digAndBuild
* enchanting

File: lib/features.json
Comment (rom1504) [inline lib/features.json]: this commit broke the tests

why are you undoing that ?

Comment (u9g) [reply lib/features.json]: theres no reason to make this change if as soon as 1.17.1 the packet is literally back to what it was

Comment (u9g) [reply lib/features.json]: fixed

Comment (rom1504) [reply lib/features.json]: can we assume no server will be running 1.17 and all 1.17.1 ?

Comment (u9g) [reply lib/features.json]: no lol, why would we just leave a bug like that

Comment (rom1504) [reply lib/features.json]: no as in we can't assume it ?
this revert introduces a bug for 1.17
we can choose not to support 1.17 and only 1.17.1 if nobody uses 1.17

Comment (u9g) [reply lib/features.json]: 1. How does it introduce a bug? 
2. Why would people skip 1.17?

Comment (rom1504) [reply lib/features.json]: 1. because 1.17 has the different entity destroy packet which you just removed support for
2. I don't know, that's what I'm asking you. If they don't skip 1.17, then this commit shouldn't be done

Comment (rom1504) [reply lib/features.json]: ah looking at the code I see you actually kept support for 1.17 , didn't read in detail before
all good then

File: lib/plugins/entities.js
Comment (rom1504) [inline lib/plugins/entities.js:37]: it's a hack to emit packet events like this, it will prevent things like mineflayer proxy from working well
let's use an event on bot instead ?

File: lib/features.json
Comment (u9g) [reply lib/features.json]: Yeah I was very confused

**Comment by jkcoxson** [comment]
The crafting, inventory and enchanting GitHub tests are likely failing from issue #2064

**Comment by nickelpro** [comment]
We came to the same conclusion on the Discord. Working on a fix

**Comment by rom1504** [comment]
we're continuing at #2106

**Pull request closed without merging**
